src/wfdir.c

```diff
@@ -645,7 +645,15 @@ DirWndProc(
       if (j == cItems)
          return -2L;
 
-      return((i + j) % cItems);
+      int pos = (i + j) % cItems;
+
+      // There is a weird behavior in listbox which selects all between anchor an caret
+      // if SHIFT is pressed. Since we return the position here and thus caret will be 
+      // updated, anchor is behind, and pressing shift selects all between anchor and caret
+      // To overcome this we select the current position, and bring anchor and cart in sync.
+      SendMessage(hwndLB, LB_SETSEL, 1, pos);

```

Comment:
(Sorry, this is just a rant.)  LB_SETSEL changes the selection state of an item; in theory, it doesn't move the caret.  However, moving the caret doesn't change the behavior here.  If we did change the selection state ourselves, that suggests we should also return -2, except if we do, the previous selection is retained (ie., `return pos` is used to deselect the previous item.)  It really looks to me like this code works because the listbox has explicit detection for LB_SETSEL which causes it to alter its behavior when this message returns.  Ick.

Also, the tree control duplicates all of this logic, but doesn't have the same problem since it's not multiselect.  Sigh.

Reply:
This change is about moving the anchor.
The caret is changed due to the return value.
We can not return -2, because we need the new position

What I saw from debugging the problem is that listbox simply selects from anchor to caret. If anchor and caret is the same we are fine. That's what we are doing with is line.

This solution is the simplest way out. It does not cost performance, It simply solves the issue and it documents the workaround.
Everything else means changing code somewhere else.

Reply:
Right, I see this is the best solution.  What's puzzling me is that this is LB_SETSEL as opposed to LB_SETCARETINDEX or LB_SETANCHORINDEX.  Changing the caret index appears to not change behavior; changing the anchor index means even though it returns pos, the new item is not selected.  So LB_SETSEL is pragmatically the right thing to do, but there's more to it than just moving the anchor.  Thinking about it today, this is probably because Shift+key selects everything from the anchor to the caret, so moving the anchor means it selects nothing, and LB_SETSEL is both moving the anchor and making the selection.  It just seems unexpected (to me at least.)